storage/repair.go

```diff
@@ -124,23 +128,280 @@ func (r shardRepairer) Repair(
 	// Add peer metadata
 	peerIter, err := session.FetchBlocksMetadataFromPeers(namespace, shard.ID(), start, end)
 	if err != nil {
-		return repair.MetadataComparisonResult{}, err
+		return repair.Result{}, err
 	}
 	if err := metadata.AddPeerMetadata(peerIter); err != nil {
-		return repair.MetadataComparisonResult{}, err
+		return repair.Result{}, err
 	}
 
 	metadataRes := metadata.Compare()
+	result := repair.Result{DifferenceSummary: metadataRes}
 
-	r.recordFn(namespace, shard, metadataRes)
+	res, err := r.repairFn(namespace, shard, metadataRes)
+	if err != nil {
+		// even in the event of repair failure, report statistics about differences
+		r.recordFn(namespace, shard, result)
+		return result, err
+	}
 
-	return metadataRes, nil
+	result.RepairSummary = res
+	r.recordFn(namespace, shard, result)
+	return result, nil
 }
 
-func (r shardRepairer) recordDifferences(
+type hostSet map[string]repair.HostBlockMetadata
+
+type blockID struct {
+	idHash ts.Hash
+	start  time.Time
+}
+
+func newHostSet() *hostSet {
+	hs := make(hostSet, 3)
```

Comment:
Don't think we should use the magic number `3` here, shouldn't it be the replica count?